### ISSUE TITLE
Java exception escaping on `Regexp#to_s`

### ISSUE THREAD

**Issue by nirvdrum** [oracle/truffleruby#2928]
I'm working on refreshing our tags for the MRI tests. While doing so, I ran into a situation where a Java exception is escaping into Ruby when trying to turn an extended regex into a string.

```ruby
re = /#\g#{"\n"}/x
re = /#{re}/
p re.to_s
```

On MRI 3.1.3:

```
> ruby -v bad-extended-regexp.rb
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [arm64-darwin22]
"(?x-mi:#\\g\n)"
```

On TruffleRuby 23.0.0-dev:

```
> jt ruby -v bad-extended-regexp.rb
truffleruby 23.0.0-dev-671182fb*, like ruby 3.1.3, GraalVM CE JVM [aarch64-darwin]

truffleruby: an internal exception escaped out of the interpreter,
please report it to https://github.com/oracle/truffleruby/issues.

offset 7, count 11, length 12 (java.lang.StringIndexOutOfBoundsException)
        from java.lang.String.checkBoundsOffCount(String.java:4587)
        from java.lang.String.<init>(String.java:523)
        from java.lang.String.<init>(String.java:1413)
        from org.joni.Lexer.syntaxWarn(Lexer.java:1327)
        from org.joni.Lexer.fetchTokenFor_subexpCall(Lexer.java:916)
        from org.joni.Lexer.fetchToken(Lexer.java:1152)
        from org.joni.Parser.parseStringLoop(Parser.java:1073)
        from org.joni.Parser.parseExpTkByte(Parser.java:1068)
        from org.joni.Parser.parseExp(Parser.java:799)
        from org.joni.Parser.parseBranch(Parser.java:1325)
        from org.joni.Parser.parseSubExp(Parser.java:1351)
        from org.joni.Parser.parseRegexp(Parser.java:1384)
        from org.joni.Analyser.compile(Analyser.java:78)
        from org.joni.Regex.<init>(Regex.java:155)
        from org.joni.Regex.<init>(Regex.java:134)
        from org.truffleruby.core.regexp.ClassicRegexp.toByteArrayBuilder(ClassicRegexp.java:905)
        from org.truffleruby.core.regexp.RegexpNodes$ToSNode.createTString(RegexpNodes.java:140)
        from org.truffleruby.core.regexp.RegexpNodesFactory$ToSNodeFactory$ToSNodeGen.executeAndSpecialize(RegexpNodesFactory.java:629)
        from org.truffleruby.core.regexp.RegexpNodesFactory$ToSNodeFactory$ToSNodeGen.execute(RegexpNodesFactory.java:606)
        from org.truffleruby.language.RubyCoreMethodRootNode.execute(RubyCoreMethodRootNode.java:58)
bad-extended-regexp.rb:3:in `to_s'
        from bad-extended-regexp.rb:3:in `<main>'
```

**Comment by eregon** [comment]
Potentially we could workaround by using `WarnCallback.NONE` until it's fixed in Joni, but maybe that introduces some other incompatibilities?
BTW you can make a PR to Joni if you'd like, but we'd need to ask for a release.
Also fine to just wait, as long as this only happens in some MRI test, we could just exclude it.

**Comment by andrykonchin** [comment]
I will prepare a PR to fix the issue in Joni.

**Comment by eregon** [comment]
It got fixed on May 22, but now it needs a Joni release and us to import a newer version of Joni

**Comment by andrykonchin** [comment]
Fixed in d484abea48b437849a4195c05fcb649ec237f6db
